File: cpp/line_test.cpp
#include "line.hpp"  // Line
#include "point.hpp" // Point
#include <cassert>   // assert
#include <iostream>  // cout
#include <sstream>   // stringstream
#include <limits>
using namespace std;

int main() {
	// No rounding error in symmetric case
	{
		auto a = Point(1001, 1);
		auto b = Point(-1000, -1);
		Line l(a, b);
		auto c = Point(0, 0);
		assert(!l.intersects(c));
		auto d = Point(3002, 3);
		assert(l.intersects(d));
		stringstream ss;
		ss << l;
		assert(ss.str() == "y = 0.0009995x + -0.00049975");
	}
	// No rounding error in asymmetric case
	{
		auto a = Point(100000, 100000);
		auto b = Point(-1, 0);
		Line l(a, b);
		auto c = Point(0, 1);
		assert(!l.intersects(c));
		auto d = Point(200001, 200000);
		assert(l.intersects(d));
		stringstream ss;
		ss << l;
		assert(ss.str() == "y = 0.99999x + 0.99999");
	}
	// Largest value with no rounding error
	{
		auto a = Point(134217726, 134217727);
		auto b = Point(-1, 134217726);
		Line l(a, b);
		auto c = Point(0, 134217726);
		assert(!l.intersects(c));
		stringstream ss;
		ss << l;
		assert(ss.str() == "y = 7.45058e-09x + 1.34218e+08");
	}
	// Smallest actual rounding error
	{
		auto a = Point(134217727, 134217728);
		auto b = Point(-1, 134217727);
		Line l(a, b);
		auto c = Point(0, 134217727);
		assert(l.intersects(c));
		stringstream ss;
		ss << l;
		assert(ss.str() == "y = 7.45058e-09x + 1.34218e+08");
	}
	// Vertical line
	{
		auto a = Point(5, 7);
		auto b = Point(5, 3);
		Line l(a, b);
		auto c = Point(5, 9);
		assert(l.intersects(c));
		stringstream ss;
		ss << l;
		assert(ss.str() == "x = 5");
	}
	// Horizontal line
	{
		auto a = Point(7, 5);
		auto b = Point(3, 5);
		Line l(a, b);
		auto c = Point(9, 5);
		assert(l.intersects(c));
		stringstream ss;
		ss << l;
		assert(ss.str() == "y = 0x + 5");
	}
	// Positive slope
	{
		auto a = Point(1, 2);
		auto b = Point(5, 4);
		Line l(a, b);
		auto c = Point(7, 5);
		assert(l.intersects(c));
		stringstream ss;
		ss << l;
		assert(ss.str() == "y = 0.5x + 1.5");
	}
	// Negative slope
	{
		auto a = Point(1, 4);
		auto b = Point(5, 2);
		Line l(a, b);
		auto c = Point(3, 3);
		assert(l.intersects(c));
		stringstream ss;
		ss << l;
		assert(ss.str() == "y = -0.5x + 4.5");
	}
	// Coprime points
	{
		auto a = Point(10, 14);
		auto b = Point(25, 35);
		Line l(a, b);
		auto c = Point(0, 0);
		assert(l.intersects(c));
		stringstream ss;
		ss << l;
		assert(ss.str() == "y = 1.4x + 0");
	}
	// Same point
	{
		auto a = Point(2, -5);
		auto b = Point(2, -5);
		Line l(a, b);
		auto c = Point(2, -5);
		assert(l.intersects(c));
		stringstream ss;
		ss << l;
		assert(ss.str() == "x = 2");
	}
	cout << "Test passed.\n";
	return 0;
}

File: dart/test/line_test.dart
import 'package:dart/line.dart';
import 'package:dart/point.dart';
import 'package:test/test.dart';

void main() {
	test('line no rounding error symmetric', () {
		final a = Point(1001, 1);
		final b = Point(-1000, -1);
		final l = Line(a, b);
		final c = Point(0, 0);
		expect(l.intersects(c), false);
		final d = Point(3002, 3);
		expect(l.intersects(d), true);
	});

	test('line no rounding error asymmetric', () {
		final a = Point(100000, 100000);
		final b = Point(-1, 0);
		final l = Line(a, b);
		final c = Point(0, 1);
		expect(l.intersects(c), false);
		final d = Point(200001, 200000);
		expect(l.intersects(d), true);
	});

	test('line largest no rounding error', () {
		final a = Point(134217726, 134217727);
		final b = Point(-1, 134217726);
		final l = Line(a, b);
		final c = Point(0, 134217726);
		expect(l.intersects(c), false);
		expect(l.toString(), 'y = 7.45058065243498e-9x + 134217726.00000001');
	});

	test('line smallest rounding error', () {
		final a = Point(134217727, 134217728);
		final b = Point(-1, 134217727);
		final l = Line(a, b);
		final c = Point(0, 134217727);
		expect(l.intersects(c), true);
		expect(l.toString(), 'y = 7.450580596923828e-9x + 134217727.0');
	});

	test('line vertical', () {
		final a = Point(5, 7);
		final b = Point(5, 3);
		final l = Line(a, b);
		final c = Point(5, 9);
		expect(l.intersects(c), true);
		expect(l.toString(), 'x = 5.0');
	});

	test('line horizontal', () {
		final a = Point(7, 5);
		final b = Point(3, 5);
		final l = Line(a, b);
		final c = Point(9, 5);
		expect(l.intersects(c), true);
		expect(l.toString(), 'y = 0.0x + 5.0');
	});

	test('line upward', () {
		final a = Point(1, 2);
		final b = Point(5, 4);
		final l = Line(a, b);
		final c = Point(7, 5);
		expect(l.intersects(c), true);
		expect(l.toString(), 'y = 0.5x + 1.5');
	});

	test('line downward', () {
		final a = Point(1, 4);
		final b = Point(5, 2);
		final l = Line(a, b);
		final c = Point(3, 3);
		expect(l.intersects(c), true);
		expect(l.toString(), 'y = -0.5x + 4.5');
	});

	test('line same points', () {
		final a = Point(2, -5);
		final b = Point(2, -5);
		final l = Line(a, b);
		final c = Point(-5, 2);
		expect(l.intersects(c), false);
		final d = Point(2, 7);
		expect(l.intersects(d), true);
		expect(l.toString(), 'x = 2.0');
	});
}
